File: java/208.java
// 208. Implement Trie (Prefix Tree)
//   Implement a trie with insert, search, and startsWith methods.
//
//   Example:
//
//   Trie trie = new Trie();
//
//   trie.insert("apple");
//   trie.search("apple");   // returns true
//   trie.search("app");     // returns false
//   trie.startsWith("app"); // returns true
//   trie.insert("app");
//   trie.search("app");     // returns true
//   Note:
//
//   You may assume that all inputs are consist of lowercase letters a-z.
//   All inputs are guaranteed to be non-empty strings.
// Runtime: 32 ms, faster than 60.44% of Java online submissions for Implement Trie (Prefix Tree).
// Memory Usage: 49.9 MB, less than 5.10% of Java online submissions for Implement Trie (Prefix Tree).

class Trie {

    Node root;
    class Node {
        boolean isEnd;
        Node[] nextLayer;
        public Node() {
            isEnd = false;
            nextLayer = new Node[26];
        }
    }
    
    /** Initialize your data structure here. */
    public Trie() {
        root = new Node();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        char[] arr = word.toCharArray();
        Node cur = root;
        for (int i = 0; i < arr.length; i++) {
            int index = arr[i] - 'a';
            if (cur.nextLayer[index] == null) {
                cur.nextLayer[index] = new Node();
            }
            cur = cur.nextLayer[index];
            if (i == arr.length - 1) {
                cur.isEnd = true;
            }
        }
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        char[] arr = word.toCharArray();
        Node cur = root;
        for (int i = 0; i < arr.length; i++) {
            int index = arr[i] - 'a';
            if (cur.nextLayer[index] == null) {
                return false;
            }
            cur = cur.nextLayer[index];
        }
        return cur.isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        char[] arr = prefix.toCharArray();
        Node cur = root;
        for (int i = 0; i < arr.length; i++) {
            int index = arr[i] - 'a';
            if (cur.nextLayer[index] == null) {
                return false;
            }
            cur = cur.nextLayer[index];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: go/208.go
// 208. Implement Trie (Prefix Tree)

// A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. There are various applications of this data structure, such as autocomplete and spellchecker.

// Implement the Trie class:

// Trie() Initializes the trie object.
// void insert(String word) Inserts the string word into the trie.
// boolean search(String word) Returns true if the string word is in the trie (i.e., was inserted before), and false otherwise.
// boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the prefix prefix, and false otherwise.
 

// Example 1:

// Input
// ["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
// [[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
// Output
// [null, null, true, false, true, null, true]

// Explanation
// Trie trie = new Trie();
// trie.insert("apple");
// trie.search("apple");   // return True
// trie.search("app");     // return False
// trie.startsWith("app"); // return True
// trie.insert("app");
// trie.search("app");     // return True
 

// Constraints:

// 1 <= word.length, prefix.length <= 2000
// word and prefix consist only of lowercase English letters.
// At most 3 * 104 calls in total will be made to insert, search, and startsWith.

// Runtime 19 ms Beats 76.66%
// Memory 18.42 MB Beats 20.58%
type Trie struct {
    NextLayer []*Trie
    HasEnd bool
}


func Constructor() Trie {
    return Trie{
        NextLayer: make([]*Trie, 26),
    }   
}


func (this *Trie) Insert(word string)  {
    cur := this
    for index, char := range(word) {
        i := char - rune('a')
        if cur.NextLayer[i] == nil {
            cur.NextLayer[i] = &Trie{NextLayer: make([]*Trie, 26)}
        }
        cur = cur.NextLayer[i]
        if index == len(word) - 1 {
            cur.HasEnd = true
        }
    }
}


func (this *Trie) Search(word string) bool {
    cur := this
    for _, char := range(word) {
        i := char - rune('a')
        if cur.NextLayer[i] == nil {
            return false
        }
        cur = cur.NextLayer[i]
    }
    return cur.HasEnd
}


func (this *Trie) StartsWith(prefix string) bool {
    cur := this
    for _, char := range(prefix) {
        i := char - rune('a')
        if cur.NextLayer[i] == nil {
            return false
        }
        cur = cur.NextLayer[i]
    }
    return true
}